src/passes/OptimizeInstructions.cpp

```diff
@@ -1738,16 +1743,99 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // (x op C1) + (x op C2)   ==>   x * (C1' + C2')
+  // (x op C1) - (x op C2)   ==>   x * (C1' - C2')
+  // where op = (`*`|`<<`)
+  // x * y + x * z   ==>   (y + z) * x
+  // x * y - x * z   ==>   (y - z) * x
+  Expression* optimizeBinaryDistributions(Binary* curr) {
+    using namespace Match;
+    assert(curr->type.isInteger());
+
+    if (curr->op == Abstract::getBinary(curr->type, Abstract::Add) ||
+        curr->op == Abstract::getBinary(curr->type, Abstract::Sub)) {
+      Expression* ll = nullptr;
+      Expression* rl = nullptr;
+      Expression* lr = nullptr;
+      Expression* rr = nullptr;
+      Binary* left = nullptr;
+      Binary* right = nullptr;
+
+      if (matches(curr->left, binary(&left, any(&ll), any(&lr))) &&
+          matches(curr->right, binary(&right, any(&rl), any(&rr))) &&
+          !effects(curr->right).hasSideEffects()) {
```

Comment:
I still think it would be easier to understand the pattern that is being matched if the match started with `curr` rather than `curr->left` and `curr->right`. Also, you can use `pure` instead of `any` to check side effects (feel free to add a new overload of `pure` if necessary`.